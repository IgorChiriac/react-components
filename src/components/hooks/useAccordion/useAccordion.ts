import { useRef, useState, Dispatch, SetStateAction, KeyboardEvent, RefObject } from 'react';
import { isArrowDown, isArrowUp } from '../../../helpers/keyboard-keys';
import { mod } from '../../../helpers/utils';

interface IAccordionHeaderProps {
  'aria-controls': string;
  'aria-disabled': boolean;
  'aria-expanded': boolean;
  id: string;
  key: string;
  onClick: () => void;
  onFocus: () => void;
  onKeyDown: (e: KeyboardEvent<HTMLButtonElement>) => void;
  ref: (node: HTMLButtonElement) => void;
}

export type TGetAccordionHeaderProps = (id: string, index: number) => IAccordionHeaderProps;

interface IAccordionSectionStyles {
  overflow: string;
  transition: string;
  height: string;
}

interface IAccordionSectionProps {
  'aria-hidden': boolean;
  'aria-labelledby': string;
  id: string;
  key: string;
  ref: (node: HTMLDivElement) => void;
  role: string;
  style: IAccordionSectionStyles;
}

export type TGetAccordionSectionProps = (id: string, index: number) => IAccordionSectionProps;

export type TIsActiveAccordionSection = (index: number) => boolean;

type TActiveSections = number[];

let activeSections: TActiveSections;
let updateActiveSections: Dispatch<SetStateAction<TActiveSections>>;

const useAccordion = () => {
  const headersRefs = useRef<RefObject<HTMLButtonElement>['current'][]>([]).current;

  const getHeaderRef = (index: number) => (node: HTMLButtonElement) => {
    if (node === null) {
      headersRefs.splice(index, 1);
    } else {
      headersRefs[index] = node;
    }
  };

  const sectionsRefs = useRef<RefObject<HTMLDivElement>['current'][]>([]).current;

  const getSectionRef = (index: number) => (node: HTMLDivElement) => {
    if (node === null) {
      sectionsRefs.splice(index, 1);
    } else {
      sectionsRefs[index] = node;
    }
  };

  [activeSections, updateActiveSections] = useState<TActiveSections>([]);

  const isActiveSection: TIsActiveAccordionSection = index => activeSections.includes(index);

  const getSectionStyle = (index: number) => {
    const sectionRef = sectionsRefs[index];
    const baseStyle = {
      overflow: 'hidden',
      transition: 'height 200ms linear',
    };
    if (!sectionRef || !isActiveSection(index)) {
      return { ...baseStyle, height: '0px' };
    }
    return { ...baseStyle, height: `${sectionRef.clientHeight}px` };
  };

  const toggleAccordionSection = (index: number) =>
    updateActiveSections(prevSections =>
      prevSections.includes(index) ? prevSections.filter(i => i !== index) : prevSections.concat(index),
    );

  const [cursorPosition, updateCursorPosition] = useState<number>(-1);

  const getOnClick = (index: number) => () => {
    updateCursorPosition(index);
    toggleAccordionSection(index);
  };

  const getOnFocus = (index: number) => () => updateCursorPosition(index);

  const getLinkingId = (id: string) => `${id}-section`;

  const focusOnHeader = (nextCursorPosition: number) => {
    const headerRef = headersRefs[nextCursorPosition];
    if (headerRef) {
      headerRef.focus();
    }
  };

  const onKeyDown = (e: KeyboardEvent<HTMLButtonElement>) => {
    const { length } = headersRefs;
    if (isArrowUp(e)) {
      e.preventDefault();
      const nextCursorPosition = mod(cursorPosition - 1, length);
      updateCursorPosition(nextCursorPosition);
      focusOnHeader(nextCursorPosition);
    } else if (isArrowDown(e)) {
      e.preventDefault();
      const nextCursorPosition = mod(cursorPosition + 1, length);
      updateCursorPosition(nextCursorPosition);
      focusOnHeader(nextCursorPosition);
    }
  };

  const getHeaderProps: TGetAccordionHeaderProps = (id, index) => ({
    'aria-controls': getLinkingId(id),
    'aria-disabled': isActiveSection(index),
    'aria-expanded': isActiveSection(index),
    id,
    key: id,
    onClick: getOnClick(index),
    onFocus: getOnFocus(index),
    onKeyDown,
    ref: getHeaderRef(index),
  });

  const getSectionProps: TGetAccordionSectionProps = (id, index) => ({
    'aria-hidden': !isActiveSection(index),
    'aria-labelledby': id,
    id: getLinkingId(id),
    key: getLinkingId(id),
    ref: getSectionRef(index),
    role: 'region',
    style: getSectionStyle(index),
  });

  return {
    getHeaderProps,
    getSectionProps,
    isActiveSection,
  };
};

export default useAccordion;
