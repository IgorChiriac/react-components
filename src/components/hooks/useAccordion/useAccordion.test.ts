import { act, testHook } from 'react-testing-library';

import useAccordion from './useAccordion';

describe('useAccordion', () => {
  let getHeaderProps;
  let getSectionProps;
  let isActiveSection;

  beforeEach(() => {
    testHook(() => {
      const accordion = useAccordion();
      getHeaderProps = accordion.getHeaderProps;
      getSectionProps = accordion.getSectionProps;
      isActiveSection = accordion.isActiveSection;
    });
  });

  it('should render correct tab and section props', () => {
    const header = getHeaderProps('id', 1);
    const section = getSectionProps('id', 1);
    expect(header).toMatchSnapshot();
    expect(section).toMatchSnapshot();
  });

  it('should open and close espective sections on header clicks', () => {
    const { onClick: firstHeaderClick } = getHeaderProps('id', 1);
    expect(isActiveSection(1)).toEqual(false);
    act(() => {
      firstHeaderClick();
    });
    expect(isActiveSection(1)).toEqual(true);

    const { onClick: secondHeaderClick } = getHeaderProps('id', 2);
    expect(isActiveSection(2)).toEqual(false);
    act(() => {
      secondHeaderClick();
    });
    expect(isActiveSection(2)).toEqual(true);
    act(() => {
      secondHeaderClick();
    });
    expect(isActiveSection(2)).toEqual(false);
  });

  it('should render matching header and section attributes', () => {
    const section = getSectionProps('another-id', 3);
    const header = getHeaderProps('another-id', 3);
    expect(header.id).toEqual(section['aria-labelledby']);
  });
});
