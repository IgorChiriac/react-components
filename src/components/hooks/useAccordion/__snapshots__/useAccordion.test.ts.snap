// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`useAccordion should render correct tab and section props 1`] = `
Object {
  "aria-controls": "id-tab",
  "aria-disabled": false,
  "aria-expanded": false,
  "id": "id",
  "key": "id",
  "onClick": [Function],
  "onFocus": [Function],
  "onKeyDown": [Function],
  "ref": [Function],
}
`;

exports[`useAccordion should render correct tab and section props 2`] = `
Object {
  "aria-hidden": true,
  "aria-labelledby": "id",
  "id": "id-tab",
  "key": "id-tab",
  "ref": [Function],
  "role": "region",
  "style": Object {
    "height": "0px",
    "overflow": "hidden",
    "transition": "height 200ms linear",
  },
}
`;
